awsshell/resource/index.py

```diff
@@ -126,71 +137,118 @@ def describe_autocomplete(self, service, operation, param):
-        self._completer_cache = {}
-        self._update_loader_paths()
+
+    def create_client(self, service_name):
+        if service_name not in self._client_cache:
+            client = self._session.create_client(service_name)
+            self._client_cache[service_name] = client
+        return self._client_cache[service_name]
+
+
+class CompleterDescriberCreator(object):
+    """Create and cache CompleterDescriber objects."""
+    def __init__(self, loader):
+        #: A botocore.loader.Loader
+        self._loader = loader
+        self._describer_cache = {}
+        self._services_with_completions = None
+
+    def create_completer_query(self, service_name):
+        """Create a CompleterDescriber for a service.
+
+        :type service_name: str
+        :param service_name: The name of the service, e.g. 'ec2'
+
+        :return: A CompleterDescriber object.
+
+        """
+        if service_name not in self._describer_cache:
+            query = self._create_completer_query(service_name)
+            self._describer_cache[service_name] = query
+        return self._describer_cache[service_name]
+
+    def _create_completer_query(self, service_name):
+        completions_model = self._loader.load_service_model(
+            service_name, 'completions-1')
+        cq = CompleterDescriber({service_name: completions_model})
+        return cq
+
+    def services_with_completions(self):
+        if self._services_with_completions is not None:
+            return self._services_with_completions
         self._services_with_completions = set(
             self._loader.list_available_services(type_name='completions-1'))
+        return self._services_with_completions
 
-    def _update_loader_paths(self):
-        completions_path = os.path.join(
-            os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
-            'data')
-        self._loader.search_paths.insert(0, completions_path)
-
-    def _get_completer_for_service(self, service_name):
-        if service_name not in self._completer_cache:
-            api_version = self._loader.determine_latest_version(
-                service_name, 'completions-1')
-            completions_model = self._loader.load_service_model(
-                service_name, 'completions-1', api_version)
-            cq = CompleterQuery({service_name: completions_model})
-            self._completer_cache[service_name] = cq
-        return self._completer_cache[service_name]
-
-    def _get_client(self, service_name):
-        if service_name in self._client_cache:
-            return self._client_cache[service_name]
-        client = self.session.client(service_name)
-        self._client_cache[service_name] = client
-        return client
+
+class ServerSideCompleter(object):
+    def __init__(self, client_creator, describer_creator):
+        # session is a boto3 session.
+        # It is a public attribute as it is intended to be
+        # changed if the profile changes.
```

Comment:
I'll remove that comment. It was for the previous version of the code and should have been removed.  Good catch.